server/mailet/mock-smtp-server/src/test/java/org/apache/james/mock/smtp/server/MockSMTPServerTest.java

```diff
@@ -184,6 +187,157 @@ void serverShouldReceiveMessageDataClient() throws Exception {
+            sendMessageIgnoreError(mail1);
+            sendMessageIgnoreError(mail1);
+
+            sendMessageIgnoreError(mail1);
+            Awaitility.await().atMost(Duration.TEN_SECONDS)
+                .untilAsserted(() -> assertThat(mockServer.listReceivedMails()).hasSize(1));
+        }
+
+        @Test
+        void serverShouldNotDecreaseNonMatchedBehavior() throws Exception {
+            int matchedBehaviorAnswers = 2;
+            MockSMTPBehavior matched = new MockSMTPBehavior(
+                MAIL_FROM,
+                Condition.MATCH_ALL,
+                Response.serverReject(SERVICE_NOT_AVAILABLE_421, "matched"),
+                MockSMTPBehavior.NumberOfAnswersPolicy.times(matchedBehaviorAnswers));
+
+            int nonMatchedBehaviorAnswers = 3;
+            MockSMTPBehavior nonMatched = new MockSMTPBehavior(
+                RCPT_TO,
+                new Condition.OperatorCondition(Operator.CONTAINS, "nonMatched"),
+                Response.serverReject(SERVICE_NOT_AVAILABLE_421, "non matched"),
+                MockSMTPBehavior.NumberOfAnswersPolicy.times(nonMatchedBehaviorAnswers));
+
+            behaviorRepository.setBehaviors(matched, nonMatched);
+
+            sendMessageIgnoreError(mail1);
+
+            assertThat(remainedAnswersOf(nonMatched))
+                .isEqualTo(nonMatchedBehaviorAnswers);
+        }
+
+        @Test
+        void serverShouldDecreaseRemainingAnswersOnlyOncePerMessage() throws Exception {
+            int firstBehaviorAnswers = 2;
+            MockSMTPBehavior matchesAnyFrom = new MockSMTPBehavior(
+                MAIL_FROM,
+                Condition.MATCH_ALL,
+                Response.serverReject(SERVICE_NOT_AVAILABLE_421, "any from will be matched"),
+                MockSMTPBehavior.NumberOfAnswersPolicy.times(firstBehaviorAnswers));
+
+            int secondBehaviorAnswers = 3;
+            MockSMTPBehavior matchesAnyRecipient = new MockSMTPBehavior(
+                RCPT_TO,
+                Condition.MATCH_ALL,
+                Response.serverReject(SERVICE_NOT_AVAILABLE_421, "any recipient will be matched"),
+                MockSMTPBehavior.NumberOfAnswersPolicy.times(secondBehaviorAnswers));
+
+            behaviorRepository.setBehaviors(matchesAnyFrom, matchesAnyRecipient);
+
+            sendMessageIgnoreError(mail1);
+
+            assertThat(remainedAnswersOf(matchesAnyFrom) + remainedAnswersOf(matchesAnyRecipient))
+                .isEqualTo(4);
+        }
+
+        private void sendMessageIgnoreError(FakeMail mail) {
+            try {
+                smtpClient.sendMessage(mail);
+            } catch (MessagingException | IOException e) {
+                // ignore error
+            }
+        }
+
+        private Integer remainedAnswersOf(MockSMTPBehavior nonMatched) {
+            return behaviorRepository
+                .getBehaviorInformation(nonMatched)
+                .remainingAnswersCounter()
+                .get();
```

Comment:
it looks dangerous

Reply:
that's test code, if it throw, then something was wrong. I just want to quickest + simplest code get the variable to assert in test methods